### ISSUE TITLE
[JIT] Performance regression with some regexs

### ISSUE THREAD

**Issue by cmb69** [PhilipHazel/pcre2#16]
I'm forwarding this from https://bugs.php.net/81424, but it seems that this is actually a PCRE2 issue. Consider the following (bad) regex:
````
/[^{};\/\n]+\{\}/
````
When run on a large string (e.g. https://pastebin.com/WVBR4f9T), with PCRE2 10.34 *JIT* this was fast; with PCRE2 10.35 and later it is more than hundred times slower.

If the regex is rewritten to use a lookbehind assertion (`/(?<![{};\/\n]+)\{\}/`), performance with the different PCRE2 versions is on par, so you may not consider this something to be fix-worthy. :)

There is no performance regression without JIT, so I wonder whether this regex isn't jitted anymore as of PCRE2 10.35.

**Comment by PhilipHazel** [comment]
Have you tried with the new Release Candidate (10.38-RC1)? which is available on GitHub? Please do so if you can. If there is still a problem, I'll assign this to the JIT maintainer.

**Comment by cmb69** [comment]
Yes, I also tried with 10.38-RC1.

**Comment by zherczeg** [comment]
It looks like this version introduced the issue: 21c40e638b902679bc4be88afb9f93f845dbf5b0

**Comment by zherczeg** [comment]
@cmb69 thank you for this report. It looks like I accidentally disabled a very important optimization. @PhilipHazel this patch fixes it, but release cycle is started so please decide whether you want it in this release or the next.

```
diff --git a/src/pcre2_jit_compile.c b/src/pcre2_jit_compile.c
index a3f7ebe..495920d 100644
--- a/src/pcre2_jit_compile.c
+++ b/src/pcre2_jit_compile.c
@@ -11228,7 +11228,7 @@ early_fail_type = (early_fail_ptr & 0x7);
 early_fail_ptr >>= 3;

 /* During recursion, these optimizations are disabled. */
-if (common->early_fail_start_ptr == 0)
+if (common->early_fail_start_ptr == 0 && common->fast_forward_bc_ptr == NULL)
   {
   early_fail_ptr = 0;
   early_fail_type = type_skip;
```

Btw I am happy we moved to git, this bug was easier to track down with it.

**Comment by PhilipHazel** [comment]
As it is a very small patch, I think we should include it in this release. Please go ahead and do the merge and update ChangeLog.

**Comment by zherczeg** [comment]
Fixed in dc5f96663597572f694147aeec3525003c351123.

**Comment by cmb69** [comment]
That was fast, thank you! I can confirm that the patch solves the issue, and I couldn't detect any regression with the PHP PCRE test suite.
